### PR TITLE
Migration: do not add a csproj to the solution if it already exists

### LINKED ISSUE #5343: Dotnet migrate has error in sln
- checkout rx master: https://github.com/Reactive-Extensions/Rx.NET
- go to \RxNet\Source
- delete global.json
- run dotnet migrate .\Rx.NET.sln

14 projects migrated
Error with sln

```
C:\dev\git\RxNet\Rx.NET\source [vs2017-3]> dotnet --version
1.0.0-rc3-004504
C:\dev\git\RxNet\Rx.NET\source [vs2017-3 +0 ~0 -1 !]> dotnet migrate .\Rx.NET.sln
Project System.Reactive.Core migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Core)
Project System.Reactive.Interfaces migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Interfaces)
Project System.Reactive.Linq migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Linq)
Project System.Reactive.PlatformServices migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.PlatformServices)
Project System.Reactive.Observable.Aliases migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Observable.Aliases)

Project System.Reactive.Providers migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Providers)
Project System.Reactive.Experimental migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Experimental)
Project System.Reactive.Windows.Threading migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Windows.Threading)
Project System.Reactive.Runtime.Remoting migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Runtime.Remoting)
Project System.Reactive.Windows.Forms migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.Windows.Forms)
Project System.Reactive.WindowsRuntime migration succeeded (C:\dev\git\RxNet\Rx.NET\source\System.Reactive.WindowsRuntime)
Project Microsoft.Reactive.Testing migration succeeded (C:\dev\git\RxNet\Rx.NET\source\Microsoft.Reactive.Testing)
Project Tests.System.Reactive migration succeeded (C:\dev\git\RxNet\Rx.NET\source\Tests.System.Reactive)
Project Tests.System.Reactive.Uwp.DeviceRunner migration succeeded (C:\dev\git\RxNet\Rx.NET\source\Tests.System.Reactive.Uwp.DeviceRunner)
Summary
Total Projects: 14
Succeeded Projects: 14
Failed Projects: 0

Solution C:\dev\git\RxNet\Rx.NET\source\Rx.NET.sln already contains project Tests.System.Reactive.Uwp.DeviceRunner\Tests.System.Reactive.Uwp.DeviceRunner.csproj.
```

### PR DESCRIPTION
**Pull request by jgoshi**
@piotrpMSFT @livarcocc @krwq @jonsequitur @piotroko 

Fixes #5343

### CODE REVIEW

File: src/dotnet/commands/dotnet-migrate/MigrateCommand.cs [hunk 1]
```diff
@@ -109,17 +109,18 @@ private void UpdateSolutionFile(MigrationReport migrationReport)
...
+            var csprojFilesToAdd = new HashSet<string>();
 
             var slnPathWithTrailingSlash = PathUtility.EnsureTrailingSlash(_slnFile.BaseDirectory);
             foreach (var report in migrationReport.ProjectMigrationReports)
             {
                 var reportPathWithTrailingSlash = PathUtility.EnsureTrailingSlash(report.ProjectDirectory);
-                var reportRelPath = Path.Combine(
-                    PathUtility.GetRelativePath(slnPathWithTrailingSlash, reportPathWithTrailingSlash),
-                    report.ProjectName + ".xproj");
+                var relReportPath = PathUtility.GetRelativePath(
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: how about naming this variable relativeReportPath?

Comment (jgoshi) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: ok

File: src/dotnet/commands/dotnet-migrate/MigrateCommand.cs [hunk 2]
```diff
@@ -132,7 +133,15 @@ private void UpdateSolutionFile(MigrationReport migrationReport)
                 }
                 else
                 {
-                    csprojFilesToAdd.Add(Path.Combine(report.ProjectDirectory, migratedProjectName));
+                    var csprojPath = Path.Combine(relReportPath, migratedProjectName);
+                    var slnAlreadyContainsMigratedCsproj = _slnFile.Projects
+                        .Where(p => p.FilePath == csprojPath)
+                        .Any();
+
+                    if (!slnAlreadyContainsMigratedCsproj)
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: Why do we need to do this check if we are using a hashset?

Comment (jgoshi) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: If a folder has a csproj already and a project.json file, then migration succeeds. But if that csproj already is referenced in the sln file, then we should not try to add it to the sln file. This is the cause of the referenced issue (trying to call dotnet sln add for a project that is already referenced in the sln file).

Comment (livarcocc) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: I know. But since you are using a HashSet, you won't end-up with two, right?

Comment (jgoshi) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: The HashSet guards against adding two (which I think is a good change). But this code protects from adding one. In the repro, we shouldn't even be trying to add one (so the HashSet doesn't guard against that).

Comment (livarcocc) [inline src/dotnet/commands/dotnet-migrate/MigrateCommand.cs]: Ah, I see now.

**Comment by jgoshi** [comment]
Actually, I don't need this change anymore. Issue https://github.com/dotnet/cli/issues/4851
states that we should fail the migration in these cases.

**Comment by livarcocc** [comment]
@jgoshi I don't think this is he case. We hit this during the MVP session on a repo that migrated successfully otherwise.

**Comment by jgoshi** [comment]
It's because behavior has changed. In this situation we used to fail. Now we succeed and say we skipped migrating the csproj file. But according to the issue https://github.com/dotnet/cli/issues/4851 this is problematic because we don't know what the csproj file contains and really should be re-migrating.

**Comment by jgoshi** [comment]
Turns out I do need this. The fix will be to re-name the existing csproj file if it has the same name as the migration output. This means migration succeeds and we will still need this code to not try to add the csproj (since it was present in the sln prior to migration).

**Comment by jgoshi** [comment]
Fixes #5343 #4851

**Comment by jgoshi** [comment]
@piotrpMSFT @livarcocc @krwq @jonsequitur @piotroko 

This is ready for review based on the decision to move the existing csproj file (with the same name as the migration output) to the backup directory.

**Pull request merged**